pykafka/rdkafka/producer.py

```diff
@@ -0,0 +1,133 @@
+import logging
+
+from pykafka.producer import Producer, CompressionType
+from pykafka.handlers import ResponseFuture
+from . import _rd_kafka
+
+
+log = logging.getLogger(__name__)
+
+
+class RdKafkaProducer(Producer):
+    """A librdkafka-backed version of pykafka.Producer
+
+    This aims to conform to the Producer interface as closely as possible.  One
+    notable difference is that it returns a future from produce(), through
+    which the user can later check if produced messages have made it to kafka.
+
+    For an overview of how configuration keys are mapped to librdkafka's, see
+    _mk_rdkafka_config_lists.
+    """
+    def start(self):
+        if not self._running:
+            brokers = ','.join(map(lambda b: ':'.join((b.host, str(b.port))),
+                                   self._cluster.brokers.values()))
+            conf, topic_conf = self._mk_rdkafka_config_lists()
+
+            self._rdk_producer = _rd_kafka.Producer()
+            self._rdk_producer.configure(conf=conf)
+            self._rdk_producer.configure(topic_conf=topic_conf)
+            self._rdk_producer.start(brokers, self._topic.name)
+            self._running = True
+
+            # TODO spawn thread to regularly poll() for delivery reports (and
+            # break if not _running)
+
+    def _produce(self, message_partition_tup):
+        # Delivery-reports will be available through this future
+        future = ResponseFuture(self._cluster.handler)
+
+        (key, msg), part_id = message_partition_tup
+        self._rdk_producer.produce(msg, key, part_id, future)
+        return future
```

Comment:
@kbourgoin @emmett9001 @rduplain can I bug you guys - and also everyone else whose listening in - for some input on this design choice in particular, before I spend too many hours hammering out the further implementation in the C code?

What this would do is make `produce()` return a future (either a `pykafka.handlers.ResponseFuture` or a backported `concurrent.futures.Future` or something else, if you prefer) and that would just evaluate to `None` if the message made it to the broker or would raise an appropriate exception from `pykafka.exceptions` if not (ie if max_retries got exceeded for some messages, and what went wrong). And for `self._synchronous=True` it could then just evaluate the future as part of `produce()`.

(I'd also volunteer to write the pullreq to make that work the same way in `pykafka.Producer` if we like it.)

Now the question: I'd like this futures-based design mostly because it is easy to reason about from a user perspective - 1 message in, 1 future out, and you can check it or ignore it, whatever suits - but I'd like your thoughts about the practicality of it in a bigger application. Any and all alternative interface suggestions very welcome.

(I'll note that `concurrent.futures` can also facilitate a [callback interface][0], if needs be...)

[0]: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback